docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,301 @@
+enhancement introduces much more versions which may (or may not) cause use of
+them.
+A possible solution here is implementing the cache as a LRU-like cache, making
+sure we evict the old entries when we reach a certain capacity.
+
+We should make sure we don't consume too much memory when fetching releases.
+It's especially critical when the user queries for the different versions, but
+might happen in other cases too.
+
+We need to think of the retention policy of versions. We don't expect to have
+versions removed from the upgrade-channels API, but just in case we should only
+add versions and not mirror exactly the given result of those API responses.
+As opposed to that, when a minor release gets EOL (end-of-life) and we remove it
+from the configuration, the service should remove all the stale versions from
+the database.
+
+## Design Details [optional]
+
+If an enhancement is complex enough, design details should be included. When not
+included, reviewers reserve the right to ask for this section to be filled in to
+enable more thoughtful discussion about the enhancement and it's impact.
+
+### Open Questions
+
+* Is the database usage a real necessity? The idea is having the periodic thread
+  updating the database, and then the HTTP server is able to filter versions from
+  the database at its leisure. If there's a good way to do it in-memory and it
+  doesn't consume too much of it, it might be better.
+* When implementing the binaries LRU cache, should we prioritize the latest
+  versions of each y-release? We expect those to be the most common ones, but
+  maybe it's better to let the LRU mechanism determine what are the common
+  versions by itself?
+* How should we limit database entries? In case of a bug in the code that writes
+  way too much version entries, can we make sure to mitigate the issue on our
+  side? (without interrupting app-sre for their help on manual database
+  intervention)
+
+### UI Impact
+
+The UI should prioritize the latest versions when displaying the different
+options. Only in need to customize the version (meaning installing an older one),
+then it requires clicking on 'Show all available versions' and searching for the
+right one.
+
+### Test Plan
+
+A simple e2e test can be made, but it's a bit of waste. The main logic
+all happens pre-installation and a little bit when preparing the installation.
+
+For subsystem we can simulate activity with the upgrade-channels API by using
+wiremock.
+
+QE test-plan should be fairly simple on the API part - just trying to install
+some arbitrary "older" version. When UI gets adjusted to first show latest
+versions is where they'll have to change the UI automation.
+
+## Drawbacks
+
+* It means QE no longer verify that each OpenShift version is installable via
+  assisted-installer.
+* We lose a bit of control over which versions we introduce. It means if
+  some existing version has a bug, we don't have a way to exclude it.
```

Comment:
This seems like something we could handle though, right?
Maybe not necessary on the first path, but excluding some list of specific releases should be easy enough.

Reply:
Right, we can easily add an exclude-list after finding out that a version has a bug
I'll update the doc